### PR TITLE
feature/system-command-line --> main

### PR DESCRIPTION
**Pull request by vlada-shubina**
PR for merging system-command-line feature branch to main

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/Commands/AliasAddCommand.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    internal class BaseAliasAddCommand : BaseCommand<AliasAddCommandArgs>
+    {
+        internal BaseAliasAddCommand(
+            ITemplateEngineHost host,
+            ITelemetryLogger logger,
+            NewCommandCallbacks callbacks,
+            string commandName)
+            : base(host, logger, callbacks, commandName, SymbolStrings.Command_AliasAdd_Description) { }
+
+        protected override Task<NewCommandStatus> ExecuteAsync(AliasAddCommandArgs args, IEngineEnvironmentSettings environmentSettings, InvocationContext context) => throw new NotImplementedException();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/AliasAddCommand.cs:38]: https://github.com/dotnet/templating/issues/3813

File: src/Microsoft.TemplateEngine.Cli/Commands/AliasCommand.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            ITelemetryLogger telemetryLogger,
+            NewCommandCallbacks callbacks)
+            : base(host, telemetryLogger, callbacks, "alias", SymbolStrings.Command_Alias_Description)
+        {
+            IsHidden = true;
+            this.Add(new AliasAddCommand(host, telemetryLogger, callbacks));
+            this.Add(new AliasShowCommand(host, telemetryLogger, callbacks));
+        }
+
+        protected override Task<NewCommandStatus> ExecuteAsync(AliasCommandArgs args, IEngineEnvironmentSettings environmentSettings, InvocationContext context) => throw new NotImplementedException();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/AliasCommand.cs:25]: https://github.com/dotnet/templating/issues/3813

File: src/Microsoft.TemplateEngine.Cli/Commands/AliasShowCommand.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    internal class BaseAliasShowCommand : BaseCommand<AliasShowCommandArgs>
+    {
+        internal BaseAliasShowCommand(
+            ITemplateEngineHost host,
+            ITelemetryLogger logger,
+            NewCommandCallbacks callbacks,
+            string commandName)
+            : base(host, logger, callbacks, commandName, SymbolStrings.Command_AliasShow_Description) { }
+
+        protected override Task<NewCommandStatus> ExecuteAsync(AliasShowCommandArgs args, IEngineEnvironmentSettings environmentSettings, InvocationContext context) => throw new NotImplementedException();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/AliasShowCommand.cs:37]: https://github.com/dotnet/templating/issues/3813

File: src/Microsoft.TemplateEngine.Cli/Commands/CommandExamples.cs [hunk 4]
```diff
@@ -3,20 +3,22 @@
 
 #nullable enable
 
-namespace Microsoft.TemplateEngine.Cli.CommandParsing
+namespace Microsoft.TemplateEngine.Cli.Commands
 {
+    //TODO: consider refactoring based on command definition
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/CommandExamples.cs:8]: https://github.com/dotnet/templating/issues/4034

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+                WriteCustomInstantiateHelp(context);
+                return;
+            }
+
+            IEngineEnvironmentSettings environmentSettings = GetEnvironmentSettingsFromArgs(instantiateCommandArgs);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider use cache only for help
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:41]: https://github.com/dotnet/templating/issues/2191: consider creating TemplatePackageManager.GetTemplates version which uses cache only for tab completion and help

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.NoMatchHandling.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                reporter.WriteLine(InvalidOptionsListToString(invalidOptionsList, matchInfos).Bold().Red());
+            }
+            else
+            {
+                var tokens = args.ParseResult.Tokens.Select(t => t.Value);
+                reporter.WriteLine(string.Format(LocalizableStrings.NoTemplatesMatchingInputParameters, string.Join(" ", tokens)).Bold().Red());
+            }
+            reporter.WriteLine();
+            //TODO: if we were not able to match the errors, print all the errors template by template.
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.NoMatchHandling.cs:150]: https://github.com/dotnet/templating/issues/4219

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                return base.GetCompletions(context, environmentSettings);
+            }
+
+            InstantiateCommandArgs args = ParseContext(context.ParseResult);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader? hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider new API to get templates only from cache (non async)
+            IReadOnlyList<ITemplateInfo> templates =
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:78]: https://github.com/dotnet/templating/issues/2191:  consider creating TemplatePackageManager.GetTemplates version which uses cache only for tab completion and help

File: src/Microsoft.TemplateEngine.Cli/Commands/ParserFactory.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Microsoft.TemplateEngine.Cli.Commands
+{
+    internal static class ParserFactory
+    {
+        internal static Parser CreateParser(Command command, bool disableHelp = false)
+        {
+            var builder = new CommandLineBuilder(command)
+                .UseParseErrorReporting()
+                //TODO: decide if it's needed to implement it; and implement if needed
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/ParserFactory.cs:18]: https://github.com/dotnet/templating/issues/2191: implement [parse] and [suggest] directives

File: src/Microsoft.TemplateEngine.Cli/TemplateInvoker.cs [hunk 9]
```diff
@@ -201,52 +196,37 @@ private async Task<New3CommandStatus> CreateTemplateAsync(ITemplateInfo template
...
                     }
 
-                    return HandlePostActions(instantiateResult, commandInput);
+                    return HandlePostActions(instantiateResult, templateArgs);
                 case CreationResultStatus.CreateFailed:
                     Reporter.Error.WriteLine(string.Format(LocalizableStrings.CreateFailed, resultTemplateName, instantiateResult.ErrorMessage).Bold().Red());
-                    return New3CommandStatus.CreateFailed;
+                    return NewCommandStatus.CreateFailed;
+                //this is unlikely case as these errors are caught on parse level now
+                //TODO: discuss if we need better handling here, then enhance core to return canonical names as array and not parse them from error message
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateInvoker.cs:209]: https://github.com/dotnet/templating/issues/4225

File: src/Microsoft.TemplateEngine.Cli/TemplateInvoker.cs [hunk 10]
```diff
@@ -201,52 +196,37 @@ private async Task<New3CommandStatus> CreateTemplateAsync(ITemplateInfo template
...
                     }
-                    return New3CommandStatus.MissingMandatoryParam;
+                    return NewCommandStatus.MissingMandatoryParam;
                 case CreationResultStatus.NotFound:
-                    Reporter.Error.WriteLine(string.Format(LocalizableStrings.MissingTemplateContentDetected, commandInput).Bold().Red());
-                    return New3CommandStatus.NotFound;
+                    Reporter.Error.WriteLine(string.Format(LocalizableStrings.MissingTemplateContentDetected, templateArgs).Bold().Red());
+                    return NewCommandStatus.NotFound;
+                //this is unlikely case as these errors are caught on parse level now, so rely on proper error message from core.
+                //TODO: discuss if we need better handling here, then enhance core to return canonical names as array and not parse them from error message
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateInvoker.cs:224]: https://github.com/dotnet/templating/issues/4225

File: src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliSearchFiltersFactory.cs [hunk 11]
```diff
@@ -72,23 +72,24 @@ internal CliSearchFiltersFactory(IReadOnlyList<IManagedTemplatePackage> installe
                                 group,
                                 groupFilters,
                                 templateFilters,
-                                CliFilters.ListTemplateParameterFilter(hostDataLoader, commandInput)));
+                                //TODO: implement it for template options matching, for now the filter just returns empty collection.
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliSearchFiltersFactory.cs:75]: https://github.com/dotnet/templating/issues/4061

File: src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs [hunk 12]
```diff
@@ -77,30 +71,31 @@ internal static async Task<New3CommandStatus> SearchForTemplateMatchesAsync(IEng
                 Reporter.Output.WriteLine(string.Format(LocalizableStrings.CliTemplateSearchCoordinator_Info_MatchesFromSource, result.Provider.Factory.DisplayName));
                 if (result.SearchHits.Any())
                 {
-                    DisplayResultsForPack(result.SearchHits, environmentSettings, commandInput, defaultLanguage);
+                    DisplayResultsForPack(result.SearchHits, environmentSettings, commandArgs, defaultLanguage);
                 }
                 else
                 {
-                    IReadOnlyDictionary<string, string?>? appliedParameterMatches = TemplateCommandInput.GetTemplateParametersFromCommand(commandInput);
+                    //TODO: implement it for template options matching
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:78]: https://github.com/dotnet/templating/issues/4061

File: src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs [hunk 13]
```diff
@@ -127,35 +122,31 @@ private static string EvaluatePackageToShow(IReadOnlyList<SearchResult> searchRe
         private static void DisplayResultsForPack(
             IReadOnlyList<(ITemplatePackageInfo PackageInfo, IReadOnlyList<ITemplateInfo> MatchedTemplates)> results,
             IEngineEnvironmentSettings environmentSettings,
-            INewCommandInput commandInput,
+            SearchCommandArgs commandArgs,
             string? defaultLanguage)
         {
-            IReadOnlyDictionary<string, string?>? appliedParameterMatches = TemplateCommandInput.GetTemplateParametersFromCommand(commandInput);
+            //TODO: implement it for template options matching
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:128]: https://github.com/dotnet/templating/issues/4061

File: src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs [hunk 14]
```diff
@@ -179,20 +170,22 @@ private static string EvaluatePackageToShow(IReadOnlyList<SearchResult> searchRe
             return templateGroupsForDisplay;
         }
 
-        private static bool ValidateCommandInput(INewCommandInput commandInput)
+        private static bool ValidateCommandInput(SearchCommandArgs commandArgs)
         {
-            if (string.IsNullOrWhiteSpace(commandInput.SearchNameCriteria) && SupportedFilters.All(filter => !filter.IsFilterSet(commandInput)) && !commandInput.RemainingParameters.Any())
+            if (string.IsNullOrWhiteSpace(commandArgs.SearchNameCriteria) && !commandArgs.AppliedFilters.Any())
+            //TODO: implement it for template options matching
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:176]: https://github.com/dotnet/templating/issues/4061

File: src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs [hunk 15]
```diff
@@ -201,6 +194,33 @@ private static bool ValidateCommandInput(INewCommandInput commandInput)
             return true;
         }
 
+        private static string GetInputParametersString(SearchCommandArgs commandArgs/*, IReadOnlyDictionary<string, string?>? templateParameters = null*/)
+        {
+            string separator = ", ";
+            IEnumerable<string> appliedFilters = commandArgs.AppliedFilters
+                .Select(filter => $"{commandArgs.GetFilterToken(filter)}='{commandArgs.GetFilterValue(filter)}'");
+
+            //TODO: implement it for template options matching
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:203]: https://github.com/dotnet/templating/issues/4061

File: src/dotnet-new3/ParserFactory.cs [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+namespace Dotnet_new3
+{
+    internal static class ParserFactory
+    {
+        internal static Parser CreateParser(Command command, bool disableHelp = false)
+        {
+            var builder = new CommandLineBuilder(command)
+            //.UseExceptionHandler(ExceptionHandler)
+            //.UseLocalizationResources(new CommandLineValidationMessages())
+            //TODO: decide if it's needed to implement it; and implement if needed
```
Comment (vlada-shubina) [inline src/dotnet-new3/ParserFactory.cs]: https://github.com/dotnet/templating/issues/2191: implement [parse] and [suggest] directives

File: src/dotnet-new3/ParserFactory.cs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+            //.UseParseDirective()
+            //.UseSuggestDirective()
+            .UseParseErrorReporting()//TODO: discuss with SDK if it is possible to use it.
```
Comment (vlada-shubina) [inline src/dotnet-new3/ParserFactory.cs]: https://github.com/dotnet/templating/issues/3814

**Review by DavidKarlas — APPROVED** [review]
All of this changes were already reviewed individually in feature branch, hence accepting as is.

**Pull request merged**